packages/financial-templates-lib/test/proxy-transaction-handler/DSProxyManager.js

```diff
@@ -0,0 +1,159 @@
+    const dsProxyAddress = dsProxyManager.getDSProxyAddress();
+
+    await testToken.mint(dsProxyAddress, toWei("1000"), { from: contractCreator });
+
+    // Deploy a tokenSender contract containing library code for the tx. the TokenSender contract enables a DSProxy to send
+    // any tokens that it might have within it's wallet. It is used to showcase how a more complex transaction would occur
+    // with the DSProxy acting in place of an EOA.
+    tokenSender = await TokenSender.new();
+
+    // Encode the send transaction. This is what will be executed within the DSProxy.
+    const sendTokenTx = tokenSender.contract.methods
+      .transferERC20(testToken.address, contractCreator, toWei("10"))
+      .encodeABI();
+
+    // Call the method on the tokenSender(library) with the sendTokenTxData. Note this is called from the perspective
+    // of the DSProxy and should send some of the DSProxy tokens.
+    const dsProxyCallReturn = await dsProxyManager.callFunctionOnExistingLibrary(tokenSender.address, sendTokenTx);
+
+    // We can get the events to double check token transferers were correct.
+    let tokenEvents = await testToken.getPastEvents("Transfer");
+
+    // The transaction hash should equal that in the transaction from the dsProxyManager.
+    assert.equal(tokenEvents[0].transactionHash, dsProxyCallReturn.transactionHash);
+
+    // The tokens should have been transferred out of the DSProxy wallet ant to the contractCreator.
+    assert.equal((await testToken.balanceOf(contractCreator)).toString(), toWei("10"));
+    assert.equal((await testToken.balanceOf(dsProxyAddress)).toString(), toWei("990"));
+    assert.isTrue(spyLogIncludes(spy, -1, "Executed function on deployed library"));
+    assert.isTrue(spyLogIncludes(spy, -1, tokenEvents[0].transactionHash)); // The transaction hash should be included.
+  });
+  it("Can send functions to DSProxy using an not deployed bytecode library", async function() {
+    // In this test we do the same execution as before (send tokens from DSProxy) but this time we dont deploy the
+    // contract library and rather do it in the same tx at the DSProxy call. This works by first deploying the library
+    // then making the call on it. Note this is still from the context of the DSProxy so msg.sender is the DSProxy address.
+    await dsProxyManager.initializeDSProxy();
+    const dsProxyAddress = dsProxyManager.getDSProxyAddress();
+
+    await testToken.mint(dsProxyAddress, toWei("1000"), { from: contractCreator });
+
+    // As before we encode the tx to send to the library.
+    const contract = new web3.eth.Contract(TokenSender.abi);
+    const callData = contract.methods.transferERC20(testToken.address, contractCreator, toWei("10")).encodeABI();
+
+    // The library also needs to code of the contract to deploy.
+    const callCode = TokenSender.bytecode;
+
+    const dsProxyCallReturn = await dsProxyManager.callFunctionOnNewlyDeployedLibrary(callCode, callData);
```

Comment:
can you explain what this call is doing. what are the on chain side effects of this?

Reply:
This one is similar to the above function except first the DSProxy will deploy the `callCode` byte code and then execute `callData` on the call code. This differs from `dsProxyManager.callFunctionOnExistingLibrary` which uses an already deployed byte code that is already on-chain. 

The steps are something like this:
1) DSProxy checks if the `callCode` is already deployed  and is in the cache. if it is, load it in. else, deploy the contract
2) execute call data on `callCode`. in this tx the "context" is the DSProxy. this is the key detail when working with proxies; the `msg.sender` is no longer the EOA that called the proxy but is now the proxy itself
3) the transaction `callData` is executed on the `callCode` contract with `msg.sender` now coming from the DSProxy due to the use of `delegatecall`.

Point 3, and the switching of `msg.sender` is the reason that DSProxy is so powerful: the contract can "act" as an EOA, while having all the power of a smart contract. the `callCode` could be as simple as sending an erc20 token (like in this example) or far far more complex like a) swapping from reserve to collateral currency, b) minting a synthetic and c) liquidating a position all in one tx.

Reply:
very interesting. I think i undestand, thanks for explanation. So just to be clear, the callCode gets persisted on chain after this call, its not like it gets cleaned up or deleted at the end of the call?

Reply:
correct. the cache can re-use it later on :)